src/hooks/useMedia.tsx

```diff
@@ -0,0 +1,151 @@
+    if (process.env.NODE_ENV !== 'production') {
+      // eslint-disable-next-line no-console
+      console.warn(
+        '`useMedia` When server side rendering, defaultState should be defined to prevent a hydration mismatches.'
+      )
+    }
+
+    return false
+  })
+
+  useEffect(() => {
+    if (features[mediaQueryString] !== undefined) {
+      setMatches(features[mediaQueryString] as boolean)
+    }
+  }, [features, mediaQueryString])
+
+  useEffect(() => {
+    // If `mediaQueryString` is present in features through `context` defer to
+    // the value present instead of checking with matchMedia
+    if (features[mediaQueryString] !== undefined) {
+      return
+    }
+
+    function listener(event: MediaQueryListEvent) {
+      setMatches(event.matches)
+    }
+
+    const mediaQueryList = window.matchMedia(mediaQueryString)
+
+    // Support fallback to `addListener` for broader browser support
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    // @ts-ignore this is not present in Safari <14
+    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
+    if (mediaQueryList.addEventListener) {
+      mediaQueryList.addEventListener('change', listener)
+    } else {
+      mediaQueryList.addListener(listener)
+    }
+
+    // Make sure the media query list is in sync with the matches state
+    setMatches(mediaQueryList.matches)
+
+    return () => {
+      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+      // @ts-ignore this is not present in Safari <14
+      // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
+      if (mediaQueryList.addEventListener) {
+        mediaQueryList.removeEventListener('change', listener)
+      } else {
+        mediaQueryList.removeListener(listener)
+      }
+    }
+  }, [features, mediaQueryString])
+
+  return matches
+}
+
+type MediaQueryFeatures = {
+  [key: string]: boolean | undefined
+}
+
+// Used to keep track of overrides to specific media query features, this should
+// be used for development and demo purposes to emulate specific features if
+// unavailable through devtools
+const MatchMediaContext = createContext<MediaQueryFeatures>({})
+
+type MatchMediaProps = {
+  children: React.ReactNode
+  features?: MediaQueryFeatures
+}
+
+const defaultFeatures = {}
+
+/**
+ * Use `MatchMedia` to emulate media conditions by passing in feature
+ * queries to the `features` prop. If a component uses `useMatchMedia` with the
+ * feature passed in to `MatchMedia` it will force its value to match what is
+ * provided to `MatchMedia`
+ *
+ * This should be used for development and documentation only in situations
+ * where devtools cannot emulate this feature
+ *
+ * @example
+ * <MatchMedia features={{ "(pointer: coarse)": true}}>
+ *   <Children />
+ * </MatchMedia>
+ */
+export function MatchMedia({children, features = defaultFeatures}: MatchMediaProps) {
```

Comment:
Nice! This will be so helpful for docs 👍

Reply:
OMG this is what I needed! 😍  THANK YOU SO MUCH! @joshblack